packages/keystatic/src/component-blocks/cloud-image-preview.tsx

```diff
@@ -2,159 +2,410 @@
 
-  const config = useConfig();
+function cleanImageData(
+  data: {
+    src: string;
+    width?: string | number;
+    height?: string | number;
+    alt?: string | number;
+  } = { src: '' }
+): ImageData {
+  return {
+    src: data.src,
+    alt: 'alt' in data && typeof data.alt === 'string' ? data.alt : '',
+    height: 'height' in data ? getDimension(data.height) : '',
+    width: 'width' in data ? getDimension(data.width) : '',
+  };
+}
+
+type ImageStatus = '' | 'loading' | 'good' | 'error';
+
+function ImageDialog({
+  image,
+  onChange,
+  onClose,
+}: {
+  image?: ImageData;
+  onChange: (data: ImageData) => void;
+  onClose: () => void;
+}) {
+  const [state, setState] = useState<ImageData>(cleanImageData(image));
+  const [status, setStatus] = useState<ImageStatus>('');
+  const [dimensions, setDimensions] = useState<ImageDimensions>(
+    cleanImageData()
+  );
+  const imageLibraryURL = useImageLibraryURL();
 
   const onPaste = (event: React.ClipboardEvent<HTMLInputElement>) => {
+    event.preventDefault();
     const text = event.clipboardData.getData('text/plain');
     try {
-      const data: unknown = JSON.parse(text);
+      const data: any = JSON.parse(text);
       if (
         typeof data === 'object' &&
         data !== null &&
         'src' in data &&
         typeof data.src === 'string'
       ) {
-        event.preventDefault();
-        props.onChange({
-          src: data.src,
-          alt: 'alt' in data && typeof data.alt === 'string' ? data.alt : '',
-          height: 'height' in data ? getDimension(data.height) : '',
-          width: 'width' in data ? getDimension(data.width) : '',
-        });
+        setState(cleanImageData(data));
         return;
       }
     } catch (err) {}
     const pattern = /^\s*!\[(.*)\]\(([a-z0-9_\-/:.]+)\)\s*$/;
     const match = text.match(pattern);
     if (match) {
-      event.preventDefault();
-      props.onChange({
-        src: match[2],
-        alt: match[1],
-        height: '',
-        width: '',
-      });
+      setState(
+        cleanImageData({
+          src: match[2],
+          alt: match[1],
+        })
+      );
       return;
     }
+    setState(cleanImageData({ src: text }));
   };
 
+  const src = state.src;
+
+  useEffect(() => {
+    if (!src) {
+      setStatus('');
+      return;
+    }
+    if (!isValidURL(src)) {
+      return;
+    }
+    setStatus('loading');
+    const img = new Image();
+    img.onload = () => {
+      const dimensions = {
+        width: img.width.toString(),
+        height: img.height.toString(),
+      };
+      setState(state => ({
+        ...state,
+        ...dimensions,
+      }));
+      setDimensions(dimensions);
+      setStatus('good');
+    };
+    img.onerror = () => {
+      setStatus('error');
+    };
+    img.src = src;
+    return () => {
+      img.onload = null;
+    };
+  }, [src]);
+
+  return (
+    <Dialog>
+      <Heading>
+        <Flex alignItems="center" gap="regular">
```

Comment:
Why have this inside the heading?

Reply:
Good catch, it's leftover code from the [example](https://keystar-ui.vercel.app/package/dialog#slots)